### ISSUE TITLE
Improve clickhouse events table schema

### ISSUE THREAD

**Issue by macobo** [PostHog/posthog#5684]
## Is your feature request related to a problem?

1. We can't update beyond 21.6 currently as our events table currently does `SAMPLE BY uuid`. 
    This does not work on clickhouse >= 21.7 due to https://github.com/ClickHouse/ClickHouse/issues/822 and https://github.com/ClickHouse/ClickHouse/pull/26256. 
    We can't e.g. `SAMPLE BY cityHash64(uuid)` because the expression needs to be in the sort key and a number. Related: https://github.com/PostHog/posthog/pull/5215
2. We should include `event` column in the sort key since it's used in most of our queries

Both of these are tragically pretty heavy changes which would require re-reading the whole schema again, so it might make sense to bundle these up.

Proposed solution:
```
ORDER BY (team_id, event, toStartOfHour(timestamp), distinct_id, cityHash64(uuid))
SAMPLE BY cityHash64(uuid) 
```

Open questions:
1. What's the difference in `ORDER BY (..., timestamp, ...)` and `ORDER BY (..., toDate(timestamp), ...)` etc? When to prefer one to the other?
2. How to handle this migration on cloud?
3. How to handle this migration on self-hosted?

cc @fuziontech @tiina303 @yakkomajuri would you mind handling this as #team-platform?

**Comment by tiina303** [comment]
> Both of these are tragically pretty heavy changes which would require re-reading the whole schema again, so it might make sense to bundle these up.

I'm not sure, generally like the idea of doing smaller more isolated things as much as possible to lower the risk.

> We should include event column in the sort key since it's used in most of our queries

Adding `event` column and it being after team_id, before date seems reasonable to me. 

> What's the difference in ORDER BY (..., timestamp, ...) and ORDER BY (..., toDate(timestamp), ...) etc? When to prefer one to the other?

I'm not sure if we should use `toStartOfHour(timestamp)`, but that's an interesting idea. Do you know how often people look at hourly vs daily vs minutely data (as we could also have it be minutely accuracy here `toStartOfMinute`). It seemed to me we mostly default to daily in insights etc, but usage could be different from our defaults. Downsides for making it more precise from https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#selecting-the-primary-key
"_A long primary key will negatively affect the insert performance and memory consumption, but extra columns in the primary key do not affect ClickHouse performance during SELECT queries._"

Not sure how much insert perf and memory consumption would be impacted, but I have noticed that if I try to load various graphs for minutely they hang so if we sorted minutely that could be better in theory, again not sure.

@fuziontech changed that here  https://github.com/PostHog/posthog/commit/97a665d27727bc38aa7018292dfb0af520b1963f so I'll refer to him on this one also.

> How to handle this migration on self-hosted?

After we've done it for cloud + testing just change the chart together for ClickHouse version and PostHog version.

Considering this blocking ClickHouse upgrades and M1 needing newer version we'll likely want to prioritize doing this in the near future.

**Comment by fuziontech** [comment]
We definitely shouldn't put `event` in front of `toStartOfHour(timestamp)` because the assumption is that we will always filter by some time period, not always `event`.

We use `toStartOfHour(timestamp)` because of how starting works on these tables. If you have too fine of a granularity of sort in an early `order by/sort by` clause the rest of the sorts won't make any difference.

Pulling from merge tree docs where the primary key is `(CounterID, Date)`:
```
  Whole data:     [---------------------------------------------]
  CounterID:      [aaaaaaaaaaaaaaaaaabbbbcdeeeeeeeeeeeeefgggggggghhhhhhhhhiiiiiiiiikllllllll]
  Date:           [1111111222222233331233211111222222333211111112122222223111112223311122333]
  Marks:           |      |      |      |      |      |      |      |      |      |      |
                  a,1    a,2    a,3    b,3    e,2    e,3    g,1    h,2    i,1    i,3    l,3
  Marks numbers:   0      1      2      3      4      5      6      7      8      9      10
```

if the `CounterID` here was super granular (no duplicates) the sorts order for every sort key beyond `CounterID` would be somewhat useless. Looking at it from how people query our data they are almost never looking at it from below an hour granularity and really they are looking at it date to date for most expensive analysis. Truncating the granularity allows the merge tree to better sort the child sort keys beyond the date.

Something else we should try is using bloom filters for how we are querying the `elements_chain` field https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#table_engine-mergetree-data_skipping-indexes
basically use this to skip a ton of rows before we regex them.

**Comment by macobo** [comment]
Great context re sort key, glad to have the reasoning written down. The trade-off I see is that we end up doing a lot of resorting in some queries.

Re event in sort key - you're right, but most queries that people really care about do filter on event (all of insights + sessions).

WDYT about adding 'event' after  toDaye(timestamp)? Not sure it would be a win though, since it might not be able to skip enough reads.

Alternatively, and I need to test this out, but maybe Projections is exactly what we need to support non-event-filtering querier.

All-in-all this gives some new ideas for data layout and tuning testing, thanks!

**Comment by fuziontech** [comment]
They are filtering by `time period` _and_ `event` though right? Even in sessions and insights? I hope they are at least. Everything should be filtered by date, then if event is also a filter the cardinality of rows will be further scoped down. 

The tl;dr on how I view the ordering of sort keys is from the most frequently -> least frequently used in a predicate/where clause.

I think adding `event` after `time period` is 💯 and totally makes sense since it's nice and crunchy in terms of granularity and does provide a ton of value from a predicate pov since we do filter on that so frequently.

Would love to see how Projections work here!

**Comment by macobo** [comment]
@fuziontech what's the latest on this?

**Comment by macobo** [comment]
We seem to have shipped a partial fix - we now order by 

```
ORDER BY (team_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))
```

This is not ideal since we still don't include event in the sort key, missing out on a huge potential optimization as almost all of our insights queries select for 1-2 events.

**Comment by macobo** [comment]
This migration failed on cloud. Taking over from @fuziontech

@yakkomajuri is handling the self-hosted side of this via async migrations (is there a main issue to link to).

Some notes:
- We aborted the migration due to disk filling up:
  - This is likely caused by failed copies
- Looking at the task, the some things that pop up:
  - We set a sharding_key on the table - likely this is causing inter-node traffic to copy data in much smaller chunks as we don't shard data intentionally when ingesting it.
  - We also copy to backup nodes - we could only replicate things there instead.
  - We set `insert_distributed_sync` which might be slowing us down

I'll try to make time to run this soon with a slightly different strategy:
- I'll copy the data within a node and let Replicated part of the merge tree take care of replicating the data. This means running 2 tasks - one per shard.
- I'll first try out my strategy on a large-ish partition

If that doesn't work, will try doing this via S3. We might be running into a bug fixed in 2021.7 where copier does not support copying with compatible schemas.

Will keep this task updated with new discoveries.

**Comment by macobo** [comment]
Some updates from attempting to run the migration on a single partition on cloud yesterday:

- `sharding_key` needs to always be set by clickhouse-copier, no escaping it
- The speed is blisteringly slow: 50k rows/s means it will take weeks to copy the dataset.
- It fails to clean up due to 

```
  ALTER TABLE posthog.swap_sharded_events_karl ATTACH PARTITION '202110' FROM posthog.swap_sharded_events_karl_piece_1                                                                                             
2022.01.17 21:21:55.656125 [ 615367 ] {} <Error> Application: Code: 359, e.displayText() = DB::Exception: Received from ch3.posthog.net:9000. DB::Exception: Table or Partition in posthog.swap_sharded_events_karl
 was not dropped.                                                                                                                                                                                                  
Reason:                                                                                                                                                                                                            
1. Size (97.42 GB) is greater than max_[table/partition]_size_to_drop (50.00 GB)                                                                                                                                   
2. File '/data/default/flags/force_drop_table' intended to force DROP doesn't exist                                                                                                                                
How to fix this:                                                                                                                                                                                                   
1. Either increase (or set to zero) max_[table/partition]_size_to_drop in server config                                                                                                                            
2. Either create forcing file /data/default/flags/force_drop_table and make sure that ClickHouse has write permission for it.                                                                                      
Example:                                                                                                                                                                                                           
sudo touch '/data/default/flags/force_drop_table' && sudo chmod 666 '/data/default/flags/force_drop_table'. Stack trace:   
```
- For some reason we have a ton of DEFAULT columns (expensive as they are copied) not only MATERIALIZED ones (which can be safely ignored).
- Cloud has `OPTIMIZE TABLE posthog.sharded_events FINAL` running right now since seemingly 20 hours ago which in turn eats resources. Who kicked it off and why? Who knows.
- We wouldn't have a good strategy for avoiding dropping events this way either as the dataset is consistently changing and copying requires dropping partitions to re-copy. Not sure what @fuziontech's original plan was here.

Next steps:
1. Figure out/fix the DEFAULT column issue
2. Increase rows/s: Try out other copying methods: https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-data-migration/ mentions trying to copy data using methods similar to what I did in https://github.com/PostHog/posthog/issues/7962 - using `clickhouse-client` and piping/multiple parallel processes.
3. Increase the max size to drop limits (temporarily) on cloud if still using copier.
4. For migration strategy w/o bringing down ingestion (once we have the tooling in place)
      a.When kicking off initial migration, save current timestamp T0. 
      b. Once all data copied, note down T1, copy data with `_timestamp` `T0 - some_delta`. 
      c. After that, create a separate kafka/mv table with a new consumer group, hook it up to the new table, 
      d. copy all data from `T1 - some_delta`, issue `OPTIMIZE TABLE`. 
      e. Once ready, verify we're in sync, swap and drop tables as needed.

**Comment by macobo** [comment]
Made some progress today in figuring out next steps:

1. Figure out/fix the DEFAULT column issue - all materialized columns are currently marked as DEFAULT. Pretty sure this was done by @fuziontech to make copier work properly?

2. Increasing rows/s. I did some benchmarking and found a solution:

- We should copy without the materialized columns and materialize them after the fact
- There are settings which help us tune the copying speed within a node. 

These settings are `set max_block_size=50000, max_insert_block_size=50000, max_threads=20, max_insert_threads=20, optimize_on_insert=0, max_execution_time=0` - with this we can achieve a copy speed of ~1.17M rows/s, which is good enough for now and > 20x faster than what we were working with previously. I opted not to use more threads to keep memory/cpu usage under control.

Previously suggested piping solution shouldn't be needed with this (and gets too complicated) and sorting slowed the query down. cc @yakkomajuri who might be interested in these measurements.

<details>
<summary>Click here for details</summary>

The table creation SQL and copy query:

```sql
CREATE TABLE posthog.karl_sharded_events_copy_test (
    `uuid` UUID,
    `event` String,
    `properties` String,
    `timestamp` DateTime64(6, 'UTC'),
    `team_id` Int64,
    `distinct_id` String,
    `elements_hash` String,
    `created_at` DateTime64(6, 'UTC'),
    `_timestamp` DateTime,
    `_offset` UInt64,
    `elements_chain` String,
    `$group_0` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_0'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
    `$group_1` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_1'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
    `$group_2` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_2'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
    `$group_3` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_3'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
    `$group_4` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_4'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), '')
)
ENGINE = ReplicatedReplacingMergeTree(
    '/clickhouse/prod/tables/{shard}/posthog.karl_sharded_events_copy_test',
    '{replica}',
    _timestamp
) PARTITION BY toYYYYMM(timestamp)
ORDER BY (team_id, toDate(timestamp), event, cityHash64(distinct_id), cityHash64(uuid))
SAMPLE BY cityHash64(distinct_id)
SETTINGS storage_policy = 'hot_to_cold'

INSERT INTO karl_sharded_events_copy_test(uuid, event, properties, timestamp, team_id, distinct_id, elements_hash, created_at, _timestamp, _offset, elements_chain) 
SELECT uuid, event, properties, timestamp, team_id, distinct_id, elements_hash, created_at, _timestamp, _offset, elements_chain 
FROM sharded_events 
WHERE toYYYYMM(timestamp) IN ('202110') 
LIMIT 50000000 
```

Benchmark results:

| | rows/s | time to copy 50M rows | copy speed |
| --- | --- | --- | --- |
| Current settings | 89k | 561s | 182 MB/s |
| `set max_block_size=100000, max_insert_block_size=100000, max_threads=10, max_insert_threads=10, optimize_on_insert=0, max_execution_time=0` | 700k | 71s | 1.48 GB/s |
| `set max_block_size=100000, max_insert_block_size=100000, max_threads=3, max_insert_threads=3, optimize_on_insert=0, max_execution_time=0` | 233k | 214s | 517 MB/s |
| `set max_block_size=100000, max_insert_block_size=100000, max_threads=20, max_insert_threads=20, optimize_on_insert=0, max_execution_time=0` | 1.14M | 43s | 2.35 GB/s |
| `set max_block_size=50000, max_insert_block_size=50000, max_threads=20, max_insert_threads=20, optimize_on_insert=0, max_execution_time=0` | 1.17M | 42s | 2.49 GB/s |
| `set max_block_size=100000, max_insert_block_size=100000, max_threads=20, max_insert_threads=20, optimize_on_insert=1, max_execution_time=0` | 990k | 50s | 2.07 GB/s |
| `set max_block_size=10000, max_insert_block_size=10000, max_threads=20, max_insert_threads=20, optimize_on_insert=0, max_execution_time=0` | 1.14M | 43s | 2.35 GB/s |
</details>

**Comment by guidoiaquinti** [comment]
Nice! Did you also check how much additional pressure do we put on Zookeeper? As it's a replicated table inserting and replicating ~1M rows/sec might not be trivial (or maybe yes, I don't know)

**Comment by macobo** [comment]
Ran two more experiments -
- Attempted to use clickhouse-copier with the new settings. It did not speed things up - it still did work at ~44k rows/s. This might be due to the extra materialized columns it's copying.
- I re-did the experiment above. It ingested ~1.3B events in 25 mins on a single node (~1M rows/s), then created the same replicated table on a different node but same shard - it replicated the dataset in <3 minutes.

Took a look at zookeeper as well in [grafana](https://metrics.posthog.com/d/SDE76m7Zzz/zookeeper-by-prometheus?orgId=1&refresh=5s&var-DS_PROMETHEUS=Prometheus&var-job=zookeeper&var-instance=All&from=now-3h&to=now) and found nothing too out of the ordinary.

Doing this in reverse (creating tables on all nodes, then inserting) seems to be slower at ~800k rows/s and extra zookeeper load.

With this we have enough information to run this migration in an expedient manner I believe. Will kick it off in a tmux session on a node on all shards and rely on replication to get the data to each node.

**Comment by macobo** [comment]
This >50% done, but still in progress: data copied and kafka catching up however I ran out of time and am heading out for vacation.

Documented the current state + next steps in https://github.com/PostHog/scratchpad/blob/main/karl/2022-02-01-events-table-migration-handoff.md and will take this on after I'm back.

We discussed handing this off with @guidoiaquinti and @lharress but decided not to do it without supervision.

**Comment by macobo** [comment]
Migration is now done on cloud :partying_face:

Some notes:
- I needed to redo the migration due to duplicates on one of the shards. The initial copying took a few hours
- After deduplicating, counts on the two tables did not match up. The cause seems to be that on shard 2 (ch-4) on Jan 29th there were duplicated events (1.4% of events on that day) that deduplicating on the new table removed. This was likely caused by some cluster operations we were doing then as due to postgresql troubles.
   - Debugged this together with @marcushyett-ph and decided to proceed with the migration as it fixed a data anomaly.

As the last step, will codify the learnings (there were a lot) in the handbook for the future. Closing this issue once that's done.
